server/http/httpclient/httpclient.go

```diff
@@ -0,0 +1,125 @@
+package httpclient
+
+import (
+	"errors"
+	"flag"
+	"net"
+	"net/http"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/metrics"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/net/publicsuffix"
+)
+
+var allowLocalhost = flag.Bool("httpclient.allow_localhost", false, "Allow HTTP requests to localhost")
+
+const maxHTTPTimeout = 60 * time.Minute
+
+type dialerControl = func(network, address string, conn syscall.RawConn) error
+
+func blockingDialerControl(allowed []*net.IPNet) dialerControl {
+	return func(network, address string, conn syscall.RawConn) error {
+		host, _, err := net.SplitHostPort(address)
+		if err != nil {
+			return err
+		}
+		ip := net.ParseIP(host)
+		for _, ipNet := range allowed {
+			if ipNet.Contains(ip) {
+				return nil
+			}
+		}
+		if ip.IsLoopback() || ip.IsPrivate() || ip.IsLinkLocalUnicast() || ip.IsLinkLocalMulticast() {
+			log.Infof("Blocked Fetch for address %s", address)
+			return errors.New("IP address not allowed")
+		}
+		return nil
+	}
+}
+
+func NewClientWithPrivateIPNets(timeout time.Duration, allowedPrivateIPNets []*net.IPNet) *http.Client {
+	dialerTimeout := timeout
+	if timeout == 0 || timeout > maxHTTPTimeout {
+		dialerTimeout = maxHTTPTimeout
+	}
+
+	inner := &http.Transport{
+		Dial: (&net.Dialer{
+			Timeout: dialerTimeout,
+			Control: blockingDialerControl(allowedPrivateIPNets),
+		}).Dial,
+		TLSHandshakeTimeout: timeout,
+		Proxy:               http.ProxyFromEnvironment,
+	}
+	tp := newMetricsTransport(inner)
+
+	return &http.Client{
+		Timeout:   timeout,
+		Transport: tp,
+	}
+}
+
+func NewClient() *http.Client {
```

Comment:
nit: the package already has the word 'client' in it so it's fine to leave out `Client` from the function name as it's repetitive. same for the function above.